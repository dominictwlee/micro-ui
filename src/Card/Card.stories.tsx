import { ReactElement } from 'react';
import { css } from '@emotion/react';
import { MdContentCopy } from 'react-icons/md';
import { Meta, Story } from '@storybook/react/types-6-0';

import Card from './Card';
import CardHeader from '../CardHeader';
import { Roundness, Spacing } from '../types';

const meta: Meta = {
  title: 'Card',
  component: Card,
};

export default meta;

// By passing using the Args format for exported stories, you can control the props for a component for reuse in a test
// https://storybook.js.org/docs/react/workflows/unit-testing
export const Default: Story = args => (
  <Card
    css={css`
      width: 200px;
      height: 200px;
    `}
    {...args}
  />
);

interface WithHeaderArgs {
  title?: string;
  spacing?: Spacing;
  roundness?: Roundness;
}

export const WithHeader: Story<WithHeaderArgs> = ({
  title,
  spacing,
  ...args
}) => {
  return (
    <Card
      css={css`
        max-width: 385px;
      `}
      {...args}
    >
      <CardHeader
        title={title}
        spacing={spacing}
        icon={<MdContentCopy size="1.5em" />}
      />
    </Card>
  );
};

WithHeader.storyName = 'With Header';
WithHeader.args = {
  title: 'Clipboard Settings',
};
