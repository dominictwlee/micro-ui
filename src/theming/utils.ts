const INVALID_CSS_ERROR = new Error('Invalid CSS value: Expects a string');

export function toCssVarNames<T>(themeObj: T, prefix = '-'): T {
  const vars: any = {};
  for (const [key, value] of Object.entries(themeObj)) {
    const varName = `${prefix}-${key}`;
    if (typeof value === 'object') {
      vars[key] = toCssVarNames(value, varName);
    } else if (typeof value === 'string') {
      vars[key] = `var(${varName})`;
    } else {
      throw INVALID_CSS_ERROR;
    }
  }
  return vars;
}

interface CssVarMap {
  [cssVarName: string]: string;
}

export function toCssVars<T>(themeObj: T, prefix = '-') {
  const vars: CssVarMap = {};

  for (const [key, value] of Object.entries(themeObj)) {
    const varName = `${prefix}-${key}`;
    if (typeof value === 'object') {
      const nestedVars = toCssVars(value, varName);
      for (const [nestedKey, nestedValue] of Object.entries(nestedVars)) {
        vars[nestedKey] = nestedValue;
      }
    } else if (typeof value === 'string') {
      vars[varName] = value;
    } else {
      throw INVALID_CSS_ERROR;
    }
  }
  return vars;
}
