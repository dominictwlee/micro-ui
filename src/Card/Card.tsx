import { PropsWithChildren } from 'react';
import { css } from '@emotion/react';

import { themeVars } from '../theming';

type Roundness = 1 | 2;

export interface CardProps {
  roundness?: Roundness;
}

const base = css`
  background-color: ${themeVars.colors.card.main};
`;

const shape = (roundness: Roundness) => css`
  border-radius: ${themeVars.shape.borderRadius[roundness]};
`;

export default function Card({
  roundness = 1,
  ...props
}: PropsWithChildren<CardProps>) {
  return <div css={[base, shape(roundness)]} {...props} />;
}
