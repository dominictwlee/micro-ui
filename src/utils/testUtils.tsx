import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from '../theming';

const Providers = ({ children }: any) => {
  return <ThemeProvider theme="light">{children}</ThemeProvider>;
};

const customRender = (ui: any, options: any = {}) =>
  render(ui, { wrapper: Providers, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
