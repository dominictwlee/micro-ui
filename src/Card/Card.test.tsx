import React from 'react';

import Card from '.';
import { render, screen } from '../utils/testUtils';

describe('Card', () => {
  test('render with children', () => {
    render(
      <Card>
        <h1>hello world</h1>
      </Card>
    );

    expect(screen.getByText('hello world')).toBeInTheDocument();
  });
});
