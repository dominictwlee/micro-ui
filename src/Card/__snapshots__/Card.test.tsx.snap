// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card render with default styles 1`] = `
<div
  class="css-1d5tp8j-base-shape-Card"
/>
`;
